### PR TITLE
Fix argument node types for method invocations in CFG

### PR DESCRIPTION
**Pull request by zcai1**
Fixed the types of `MethodInvocationNode#arguments` and `ObjectCreationNode#arguments` in CFG. 

Previously, argument nodes are created using the types from method declaration, which means some nodes are using type variables that are not substituted by type arguments at the call site.

For example, we used to observe `new T[]{"a", "b"}` instead of `new String[]{"a", "b"}`, while the second one makes more sense.

This PR tries to fix this issue by using the method type at the call site for argument coversions.

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 1]
```diff
@@ -1224,22 +1225,25 @@ protected Node methodInvocationConvert(Node node, TypeMirror formalType) {
...
     /**
-     * Given a method element and as list of argument expressions, return a list of {@link Node}s
-     * representing the arguments converted for a call of the method. This method applies to both
-     * method invocations and constructor calls.
+     * Given a method element, its type at the call site, and as list of argument expressions,
+     * return a list of {@link Node}s representing the arguments converted for a call of the method.
+     * This method applies to both method invocations and constructor calls.
      *
      * @param method an ExecutableElement representing a method to be called
+     * @param methodType an ExecutableType representing the type of method call
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:1233]:
```suggestion
     * @param methodType an ExecutableType representing the type of the method call
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 2]
```diff
@@ -1271,10 +1275,24 @@ protected List<Node> convertCallArguments(
...
-                    convertedNodes.add(methodInvocationConvert(actualVal, formals.get(i).asType()));
+                    convertedNodes.add(methodInvocationConvert(actualVal, formals.get(i)));
                 }
 
+                // NOTE: When the last parameter is a type variable vararg and the compiler
+                // cannot find a specific type use to substitute for it, the compiler will
+                // create an unbounded component type instead. For example,
+                //
+                // for the following method declaration:
+                // <T> void foo(T... ts) {}
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:1286]:
```suggestion
                // for the following method declaration:
                // <T> void foo(T... ts) {}
```

File: docs/CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,13 @@ Version 3.21.4-eisop2 (April ?, 2022)
 
 **Implementation details:**
 
+Fixed the types of `MethodInvocationNode#arguments` and
+`ObjectCreationNode#arguments` in CFG. Previously, argument nodes are created
```
Comment (wmdietl) [inline docs/CHANGELOG.md:9]:
```suggestion
`ObjectCreationNode#arguments` in CFGs. Previously, argument nodes are created
```

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,13 @@ Version 3.21.4-eisop2 (April ?, 2022)
...
+using the types from method declaration, which means some nodes are using type
```
Comment (wmdietl) [inline docs/CHANGELOG.md:10]:
```suggestion
using the types from the method declaration, which means some nodes are using type
```

File: framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java [hunk 5]
```diff
@@ -282,7 +282,13 @@ private AnnotatedTypeVariable getTypeVariableFromDeclaration(
...
-            if (meth == null) {
+            if (meth == null || (idx == -1 && exElt.isVarArgs())) {
+                // idx can be -1 when we have a type variable vararg in the parameters and
+                // the method invocation doesn't provide enough information about the type
+                // to substitute for it. In this situation, the `tpe` is an unbounded symbol
+                // created by the compiler, and it doesn't match any type parameters in `exElt`.
+                // Please check `CFGTranslationPhaseOne#convertCallArguments` for details.
+
                 // throw new BugInCF("TypeFromTree.forTypeVariable: did not find source for: "
                 //                   + elt);
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java:293]:
```suggestion
```

File: framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java [hunk 6]
```diff
@@ -282,7 +282,13 @@ private AnnotatedTypeVariable getTypeVariableFromDeclaration(
             ExecutableElement exElt = (ExecutableElement) elt;
             int idx = exElt.getTypeParameters().indexOf(tpe);
             MethodTree meth = (MethodTree) f.declarationFromElement(exElt);
-            if (meth == null) {
+            if (meth == null || (idx == -1 && exElt.isVarArgs())) {
+                // idx can be -1 when we have a type variable vararg in the parameters and
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java:286]:
```suggestion
            if (meth == null || (idx == -1 && exElt.isVarArgs())) {
                // meth can be null when no source code was found for it.
                // idx can be -1 when we have a type variable vararg in the parameters and
```

File: javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java [hunk 7]
```diff
@@ -505,6 +506,22 @@ public static ExecutableElement constructor(NewClassTree tree) {
         }
     }
 
+    /**
+     * Determines the type for a constructor at its call site given an invocation via {@code new}.
+     *
+     * @param tree the constructor invocation0
```
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java:512]:
```suggestion
     * @param tree the constructor invocation
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java [hunk 8]
```diff
@@ -1446,7 +1464,8 @@ public MethodInvocationNode visitMethodInvocation(MethodInvocationTree tree, Voi
             // See also BaseTypeVisitor.visitMethodInvocation and QualifierPolymorphism.annotate.
             arguments = Collections.emptyList();
         } else {
-            arguments = convertCallArguments(method, actualExprs);
+            ExecutableType methodType = (ExecutableType) TreeUtils.typeOf(tree.getMethodSelect());
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java:1467]: Should this extracted into a new `TreeUtils.typeFromUse(tree)` method, modeled after `TreeUtils.elementFromUse(tree)`?

File: javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java [hunk 9]
```diff
@@ -505,6 +506,22 @@ public static ExecutableElement constructor(NewClassTree tree) {
...
+     * @return the {@link ExecutableType} corresponding to the constructor call (i.e., the given
+     *     {@code tree}) at its call site
+     */
+    public static ExecutableType constructorType(NewClassTree tree) {
```
Comment (wmdietl) [inline javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java:516]: For a `NewClassTree` we have a `elementFromUse` and a `constructor` method that do slightly different things.
Should this method be `typeFromUse` or `constructorType`? Or do we need both?
Can you clarify in the documentation to which element this corresponds?

**Review by wmdietl — COMMENTED** [review]
Thanks! Just a few questions and small edits.

**Review by wmdietl — APPROVED** [review]
Thanks!

**Comment by wmdietl** [comment]
@zcai1 Can you please run reformat? The one character change I made unfortunately broke it.

**Comment by zcai1** [comment]
> @zcai1 Can you please run reformat? The one character change I made unfortunately broke it.

I've tried, but it didn't produce any changes and `checkFormat` passed on master.

**Comment by zcai1** [comment]
> > > @zcai1 Can you please run reformat? The one character change I made unfortunately broke it.
> > 
> > 
> > I've tried, but it didn't produce any changes and `checkFormat` passed on master.
> 
> That's odd, here's the CI error: https://dev.azure.com/eisop/checker-framework/_build/results?buildId=1091&view=logs&j=4b19ccca-7d7d-5749-26de-b87349fba2b4&t=4dba1f6c-8e7b-5172-04db-4dd554c55b12&l=451 Did you pull in my change?

Ah, apparently that was the issue. Now the reformatting should be correct.

**Pull request merged**